### PR TITLE
Make v2-test succeed when no test suites found

### PR DESCRIPTION
**Pull request by m-renaud**
# Overview 

Change the default behaviour of `v2-test` so that it succeeds when no test suites are found.

This changes the exit code behaviour of the `v2-test` command when the "problem" is that no tests were found. Previously, when no tests were found `cabal v2-test` exit with status `1`, now prints a it friendlier message and exits with status `0`. The old behaviour can be achieved by passing `--test-fail-when-no-test-suites` to `cabal v2-test`.

This addresses #5435.

## Example Usage

**--help output**
```
$ cabal v2-test help
<snip>
    --test-fail-when-no-test-suites                      Exit with failure when no
                                                         test suites are found.
<snip>
```

**Before**
```
$ cabal v2-test
Resolving dependencies...
cabal: Cannot run tests for the target '' which refers to the package
initsample-0.1.0.0 because it does not contain any test suites.

$ echo $?
1
```

**After**
```
$ cabal-dev v2-test
No tests to run for the package initsample-0.1.0.0.

$ echo $?
0
```

```
$ cabal v2-test --test-fail-when-no-test-suites
Resolving dependencies...
cabal: Cannot run tests for the target '' which refers to the package
initsample-0.1.0.0 because it does not contain any test suites.

$ echo $?
1
```

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

/cc @vrom911 @hvr

### CODE REVIEW

**Comment by fgaz** [comment]
I'd prefer the opposite: https://github.com/haskell/cabal/issues/5435#issuecomment-444109546

**Comment by 23Skidoo** [comment]
I'm +1 on @fgaz's suggestion.

**Comment by m-renaud** [comment]
@23Skidoo @fgaz see my comment on the other issue, having a double negative flag with a default value of True is quite a bit less clean IMHO.

Unless you're implying you'd be in favour of changing the default behaviour to success when no tests exist. I would actually prefer that but wasn't sure if changing this default behaviour world be breaking.

Let me know what you think.

**Comment by m-renaud** [comment]
@23Skidoo gotcha, just wanted to clarify first :) I wrote the first version with the constant that default behaviour stayed the same, although I suppose this will go out with cabal 3.0 which has a host of other breaking changes.

**Comment by fgaz** [comment]
fwiw old-test succeeded on no tests, so this isn't really breaking

**Comment by m-renaud** [comment]
Ah yes, you're right.

So, we have a couple name options, let's pick one before I change the code then have to change it again ;P

- `--test-disallow-no-test-suites`
- `--test-fail-no-test-suites`
- `--test-fail-when-no-test-suites`

I lean towards the last one.

**Comment by fgaz** [comment]
+1 for `--test-fail-when-no-test-suites`

**Comment by 23Skidoo** [comment]
`fail-no` or `fail-when-no` are both good IMO.

**Comment by m-renaud** [comment]
PR updated, CI still running but I've tested it locally and everything seems to work fine :)

File: cabal-install/Distribution/Client/CmdTest.hs [hunk 1]
```diff
@@ -206,9 +210,25 @@ data TargetProblem =
    | TargetProblemIsSubComponent   PackageId ComponentName SubComponentTarget
   deriving (Eq, Show)
 
-reportTargetProblems :: Verbosity -> [TargetProblem] -> IO a
-reportTargetProblems verbosity =
-    die' verbosity . unlines . map renderTargetProblem
+reportTargetProblems :: Flag Bool -> Verbosity -> [TargetProblem] -> IO a
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/CmdTest.hs]: `Verbosity` normally always comes first unless there are special reasons.

**Comment by 23Skidoo** [comment]
Merged, thanks!

File: cabal-install/Distribution/Client/CmdTest.hs (hunk 1 above)
Comment (m-renaud) [reply cabal-install/Distribution/Client/CmdTest.hs]: Good point, I'll send a follow-up to clean that up.

**Pull request merged**